aioredis/locks.py

```diff
@@ -0,0 +1,68 @@
+import sys
+
+from asyncio.locks import Lock as _Lock
+from asyncio import coroutine
+from asyncio import futures
+
+# Fixes an issue with all Python versions that leaves pending waiters
+# without being awakened when the first waiter is canceled.
+# Code adapted from the PR https://github.com/python/cpython/pull/1031
+# Waiting once it is merged to make a proper condition to relay on
+# the stdlib implementation or this one patched
+
+
+class Lock(_Lock):
+
+    if sys.version_info < (3, 5):  # pragma: no cover
+        @coroutine
+        def acquire(self):
+            """Acquire a lock.
+
+            This method blocks until the lock is unlocked, then sets it to
+            locked and returns True.
+            """
+            if not self._waiters and not self._locked:
+                self._locked = True
+                return True
+
+            fut = futures.Future(loop=self._loop)
+            self._waiters.append(fut)
+            try:
+                yield from fut
+                self._locked = True
+                return True
+            except futures.CancelledError:
+                if not self._locked:
+                    self._wake_up_first()
+                raise
+            finally:
+                self._waiters.remove(fut)
+    else:
```

Comment:
yes, a shame of me. Take a look to the last commit, unified to a unique function that should be compatible with all Python versions and free of issues.